pleskdistup/common/src/util.py

```diff
@@ -9,31 +9,62 @@
 # Returns standard output
 def logged_check_call(cmd: typing.Union[typing.Sequence[str], str], **kwargs) -> str:
     log.info(f"Running: {cmd!r}. Output:")
+    stdout = []
+
+    def proc_stdout(line: str) -> None:
+        stdout.append(line)
+
+    def proc_stderr(line: str) -> None:
+        log.err(line)
+
+    exit_code = exec_get_output_streamed(cmd, proc_stdout, proc_stderr, **kwargs)
+    if exit_code != 0:
+        log.err(f"Command {cmd!r} failed with return code {exit_code}")
+        raise subprocess.CalledProcessError(returncode=exit_code, cmd=cmd, output="\n".join(stdout))
+
+    log.info(f"Command {cmd!r} finished successfully")
+    return "\n".join(stdout)
 
-    # I beleive we should be able pass argument to the subprocess function
+
+# Allows to get stdout/stderr by streaming line by line, calling callbacks
+# and returns process exit code
+def exec_get_output_streamed(
+    cmd: typing.Union[typing.Sequence[str], str],
+    process_stdout_line: typing.Optional[typing.Callable[[str], None]],
+    process_stderr_line: typing.Optional[typing.Callable[[str], None]],
+    **kwargs,
+) -> int:
+    # I believe we should be able pass argument to the subprocess function
     # from the caller. So we have to inject stdout/stderr/universal_newlines
-    kwargs["stdout"] = subprocess.PIPE
-    kwargs["stderr"] = subprocess.STDOUT
+    if process_stdout_line is None:
+        kwargs["stdout"] = subprocess.DEVNULL
+    else:
+        kwargs["stdout"] = subprocess.PIPE
+    if process_stderr_line is None:
+        kwargs["stderr"] = subprocess.DEVNULL
+    else:
+        kwargs["stderr"] = subprocess.PIPE
     kwargs["universal_newlines"] = True
 
-    stdout = []
     process = subprocess.Popen(cmd, **kwargs)
-    while None is process.poll():
-        if not process.stdout:
-            log.err(f"Can't get process output from {cmd!r}")
-            raise RuntimeError(f"Can't get process output from {cmd!r}")
-        line = process.stdout.readline()
-        if line:
-            stdout.append(line)
-            if line.strip():
-                log.info(line.strip(), to_stream=False)
-
-    if process.returncode != 0:
-        log.err(f"Command {cmd!r} failed with return code {process.returncode}")
-        raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd, output="\n".join(stdout))
+    if process_stdout_line is None and process_stderr_line is None:
+        process.communicate()
+        return process.returncode
 
-    log.info(f"Command {cmd!r} finished successfully")
-    return "\n".join(stdout)
+    while None is process.poll():
+        if process_stdout_line is not None:
+            if not process.stdout:
+                raise RuntimeError(f"Cannot get process stdout of command {cmd!r}")
+            line = process.stdout.readline()
+            if line:
+                process_stdout_line(line.rstrip())
```

Comment:
Why do modify the line (remove all whitespace at the end)?

Reply:
Yes it was a tricky decision that I had to take. For the "purity" of the API, we should not remove any byte from the stream but this means that removing the end newline at the caller side. I chose to simplify caller function especially if it log.xxx(line).
If you think it should be the opposite and caller should know that it'll receive the line as is (with newline character) then it'll be OK for me although it'll add one more strip() line in callback fun.

Reply:
Why not `.rstrip('\n')` instead?

Reply:
Done.